pil-analyzer/src/pil_analyzer.rs

```diff
@@ -505,6 +505,58 @@ namespace N(16);
         let expected = r#"namespace N(16);
     let w = (|| 2);
     constant x = (|i| (|| N.w()))(2)();
+"#;
+        let formatted = analyze_string::<GoldilocksField>(input).to_string();
+        assert_eq!(formatted, expected);
+    }
+
+    #[test]
+    fn simple_type_resolution() {
+        let input = r#"namespace N(16);
+    let w: col[3 + 4];
+    "#;
+        let expected = r#"namespace N(16);
+    col witness w[7];
+"#;
+        let formatted = analyze_string::<GoldilocksField>(input).to_string();
+        assert_eq!(formatted, expected);
+    }
+
+    #[test]
+    fn complex_type_resolution() {
+        let input = r#"namespace N(16);
+    let f: int -> int = |i| i + 10;
+    let x: (int -> int), int -> int = |k, i| k(2**i);
+    let y: (int -> fe)[x(f, 2)];
+    let z: (((int -> int), int -> int)[x(|i| i, 3)], col) = ([x, x, x], y[0]);
```

Comment:
so an array is also a function?

Reply:
no, it should not. will recheck the example. This test is just about parsing the type names, evaluating the array legths and printing it again. the values are not touched (except by the array length evaluator).